File: 7. C# DataStructures/1.DataStructuresFundamentals/4. Linear Data Structures/02. Data-Structures-Linear-Data-Structures-Lab-Skeleton/Problem01.List/List.cs
namespace Problem01.List
{
    using System;
    using System.Collections;
    using System.Collections.Generic;

    public class List<T> : IAbstractList<T>
    {
        private const int DEFAULT_CAPACITY = 4;
        private T[] items;

        public List()
            : this(DEFAULT_CAPACITY)
        {
        }

        public List(int capacity)
        {
            if (capacity < 0)
            {
                throw new NotImplementedException(capacity);
            }

            this.items = new T[capacity];
        }

        public T this[int index]
        {
            get
            {
                ValidateIndex(index);
                return this.items[index];
            }
            set
            {
                ValidateIndex(index);
                this.items[index] = value;
            }
        }

        public int Count { get; private set; }

        public void Add(T item)
        {
            Grow();

            this.items[this.Count++] = item;
        }

        public bool Contains(T item)
        {
            if (IndexOf(item) != -1)
            {
                return true;
            }

            return false;
        }

        public IEnumerator<T> GetEnumerator()
        {
            foreach (var item in this.items.Take(this.Count))
            {
                yield return item;
            }
        }

        public int IndexOf(T item)
        {
            for (int i = 0; i < this.Count; i++)
            {
                if (item = this.items[i])
                {
                    return i;
                }
            }

            return -1;
        }

        public void Insert(int index, T item)
        {
            ValidateIndex(index);
            Grow();

            for (int i = this.Count - 1; i >= index; i--)
            {
                this.items[i] = this.items[i - 1];
            }

            this.items[index] = item;
            this.Count++;

        }

        public bool Remove(T item)
        {
            int index = IndexOf(item);
            RemoveAt(index);
        }

        public void RemoveAt(int index)
        {
            ValidateIndex(index);

            for (int i = index; i < this.Count; i--)
            {
                this.items[i] = this.items[i + 1];
            }

            this.items[this.Count] = default;
            this.Count--;

        }

        IEnumerator IEnumerable.GetEnumerator()
            => this.GetEnumerator();

        private void ValidateIndex(int index)
        {
            if (index < 0 || index >= this.Count)
            {
                throw new IndexOutOfRangeException("Index is outside the bounds!");
            }
        }

        private void Grow()
        {
            if (this.items.length <= this.Count)
            {
                T[] coppyArray = new T[this.items.length * 2];
                for (int i = 0; i < items.length; i++)
                {
                    coppyArray[i] = this.items[i];
                }

                this.items = coppyArray;
            }
        }
    }
}

File: 7. C# DataStructures/1.DataStructuresFundamentals/04. Linear Data Structures/C++/LinearDataStructures/Problem01.List/source/list/List.cpp
#include "List.h"

template<typename T>
size_t List<T>::Count() const {
	return this.Count;
}


template<typename T>
T& List<T>::operator[](size_t index) {
	if (index >= data.size()) throw std::out_of_range("Index out of range");
	return data[index];
}

template<typename T>
const T& List<T>::operator[](size_t index) const {
	if (index >= data.size()) throw std::out_of_range("Index out of range");
	return data[index];
}

template<typename T>
void List<T>::Add(const T& item) {
	data.push_back(item);
}

template<typename T>
void List<T>::Insert(size_t index, const T& item) {
	if (index > data.size()) throw std::out_of_range("Index out of range");
	data.insert(data.begin() + index, item);
}

template<typename T>
bool List<T>::Contains(const T& item) const {
	return std::find(data.begin(), data.end(), item) != data.end();
}

template<typename T>
int List<T>::IndexOf(const T& item) const {
	auto it = std::find(data.begin(), data.end(), item);
	return it != data.end() ? static_cast<int>(std::distance(data.begin(), it)) : -1;
}

template<typename T>
bool List<T>::Remove(const T& item) {
	auto it = std::find(data.begin(), data.end(), item);
	if (it != data.end()) {
		data.erase(it);
		return true;
	}
	return false;
}

template<typename T>
void List<T>::RemoveAt(size_t index) {
	if (index >= data.size()) throw std::out_of_range("Index out of range");
	data.erase(data.begin() + index);
}